enhancements/network/persistent-ips-sdn-secondary-networks.md

```diff
@@ -0,0 +1,948 @@
+  allocation when processing the CNI delete.
+- retrieve any `IPAMLease`s previously allocated for a Virtual Machine
+  when processing Virtual Machine's pod creations.
+  I.e. when starting a previously stopped VM, OVN-Kubernetes should check if
+  there are existing `IPAMLease`s for it, and re-use them if found.
+
+Deleting the  `IPAMLease` from the datastore when the pod is deleted will be
+performed by Kubernetes (garbage collection) once the Virtual Machine
+(i.e. owner) is removed. In the example above, that `IPAMLease` will be
+garbage collected by Kubernetes when the `VirtualMachine` named `vm-a` with UID
+`a0790345-4e84-4257-837a-e3d762d191ab` is deleted.
+
+### Implementation Details/Notes/Constraints
+
+#### Instructing the IPAM CNI the ADD/DEL requests are on behalf of an owner
+
+We require different behaviors from the IPAM CNI when processing **ADD** events:
+- when an owner is specified, the IPAM CNI plugin **will** try to recover
+  previous allocations for this particular owner. If found, they are re-used.
+  Generated otherwise.
+- when the setup command features an owner, we must persist the IP allocation
+  in the datastore for every IP allocation we generate.
+
+The behavior is also different when processing CNI deletes: the `IPAMLease`
+CRs will not be deleted by the CNI plugin: Kubernetes will garbage collect
+these once the encapsulating object is removed.
+
+The question now becomes how does the client instruct the IPAM CNI plugin this
+particular operation is for a pod owned by another entity ?
+
+#### Specifying persistent IPs for a NAD
+
+The network administrator has a say when specifying the NAD if the persistent
+IPs feature can be used for their network.
+
+To achieve that, a new flag is added (only `layer2` and `localnet` topologies),
+named `allowPersistentIPs`.
+
+```yaml
+---
+apiVersion: "k8s.cni.cncf.io/v1"
+kind: NetworkAttachmentDefinition
+metadata:
+  name: tenantred
+  namespace: default
+spec:
+  config: '{
+      "cniVersion": "0.3.1",
+      "name": "l2-network",
+      "type": "ovn-k8s-cni-overlay",
+      "topology": "layer2",
+      "mtu": 1300,
+      "netAttachDefName": "default/tenantred",
+      "subnet": "10.128.20.0/24,fd10:128:20::0/64",
+      "excludeSubnets": "10.128.200.1/32,fd10:128:20::1/128",
+      "allowPersistentIPs": true
+    }'
+```
+
+### Risks and Mitigations
+
+When building the OVN-Kubernetes IPAM module IP pools, it will be needed to
+take into account these persistent IPs as well. Thus, there will be a resource
+to sync when adding a pod requesting an attachment to a network allowing
+persistent IPs.
+
+Using informers to know the `IPAMLease`s CRs will mitigate this risk, since
+we will not actively synchronously call the Kubernetes API for reading.
+
+### Drawbacks
+
+Using the proposed solution would lead to an issue with VirtualMachine
+foreground deletion. It is described in-depth in the
+[Garbage Collection section](#kubernetes-garbage-collection-extended-info).
```

Comment:
We have to mention here that this can be fixed adding finalizer at IPAMLease, ovn-k can add it and delete it.

Reply:
OVN-K doesn't know the VM lifecycle, thus it doesn't know when to remove the finalizer. 

The finalizer option IMO is only possible if it is the user - or KubeVirt - creating the IPAM claim.